File: go/14.go
package main

import "sort"

func longestCommonPrefix(strs []string) string {
	if len(strs) == 1 { // handle only 1 element
		return strs[0]
	}

	// sort them first, the most different one will be in first and last
	sort.Strings(strs)

	// compare first and last
	l := len(strs)
	for i := range strs[0] {
		if strs[0][i] != strs[l-1][i] {
			return strs[0][:i]
		}
	}
	return strs[0]
}

/*
My thought(?), sorry for my poor English writing skill.
https://leetcode.com/problems/longest-common-prefix/solutions/3539074/go-solution-beats-100-runtime-56-memory/
*/

File: python/14.py
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        strs.sort()

        for i in range(len(strs[0])):
            if (strs[0][i] != strs[-1][i]):
                return strs[0][:i]

        return strs[0]
